File: app/src/main/java/com/sample/sv/fileprovidersample/gallery/ImageGalleryActivity.java
package com.sample.sv.fileprovidersample.gallery;

import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;

import com.sample.sv.fileprovidersample.R;

import java.util.ArrayList;
import java.util.List;

import static android.widget.GridLayout.VERTICAL;

public class ImageGalleryActivity extends AppCompatActivity {


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_gallery);

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new StaggeredGridLayoutManager(2, VERTICAL));
        ImageGalleryAdapter adapter = new ImageGalleryAdapter();
        recyclerView.setAdapter(adapter);
        adapter.setPhotos(getImagesPath());
    }

    public List<String> getImagesPath() {
        ArrayList<String> images = new ArrayList<>();
        Uri uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;

        String[] projection = {MediaStore.MediaColumns.DATA, MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
                MediaStore.Images.Media.WIDTH, MediaStore.Images.Media.HEIGHT};

        Cursor cursor = getContentResolver().query(uri, projection, null, null,
                MediaStore.MediaColumns.DATE_ADDED + " DESC");

        if (cursor != null) {
            int columnIndexData = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA);
            int columnIndexWidth = cursor.getColumnIndex(MediaStore.MediaColumns.WIDTH);
            int columnIndexHeight= cursor.getColumnIndex(MediaStore.MediaColumns.HEIGHT);
            while (cursor.moveToNext()) {
                int width = cursor.getInt(columnIndexWidth);
                int height = cursor.getInt(columnIndexHeight);
                if (width > 0 && height > 0) {
                    String absolutePathOfImage = cursor.getString(columnIndexData);
                    images.add(absolutePathOfImage);
                }
            }
            cursor.close();
        }
        return images;
    }
}

File: app/src/main/java/com/sample/sv/fileprovidersample/gallery/ImageGalleryActivity.kt
package com.sample.sv.fileprovidersample.gallery

import android.os.Bundle
import android.provider.MediaStore
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.widget.GridLayout.VERTICAL
import com.sample.sv.fileprovidersample.R
import kotlinx.android.synthetic.main.activity_gallery.*
import java.util.*

class ImageGalleryActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_gallery)

        recycler_view.layoutManager = StaggeredGridLayoutManager(2, VERTICAL)
        val adapter = ImageGalleryAdapter()
        recycler_view.adapter = adapter
        adapter.setPhotos(getImagesPath())
    }

    private fun getImagesPath(): List<String> {
        val images = mutableListOf<String>()
        val uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI

        val projection = arrayOf(
                MediaStore.MediaColumns.DATA,
                MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
                MediaStore.Images.Media.WIDTH,
                MediaStore.Images.Media.HEIGHT
        )

        val cursor = contentResolver.query(
                uri,
                projection,
                null,
                null,
                MediaStore.MediaColumns.DATE_ADDED + " DESC"
        )

        if (cursor != null) {
            val columnIndexData = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATA)
            val columnIndexWidth = cursor.getColumnIndex(MediaStore.MediaColumns.WIDTH)
            val columnIndexHeight = cursor.getColumnIndex(MediaStore.MediaColumns.HEIGHT)
            while (cursor.moveToNext()) {
                val width = cursor.getInt(columnIndexWidth)
                val height = cursor.getInt(columnIndexHeight)
                if (width > 0 && height > 0) {
                    val absolutePathOfImage = cursor.getString(columnIndexData)
                    images.add(absolutePathOfImage)
                }
            }
            cursor.close()
        }
        return images
    }
}
